EX2Linux.cpp

```diff
@@ -56,58 +67,142 @@ void freeArgs(std::vector<char*>& args) {
+ * The child process now checks if the args vector consists of any of these operators '<' '>'
+ * '|' and accordingly uses the required system calls to fullfil the requirements if the found operator.
+ */
+
 void executeCommand(const std::string& command, std::vector<char*>& args, bool isBackground,
-                    std::vector<pid_t>& backgroundProcesses) {
-    pid_t pid = fork();
-    if(pid == 0){
-        //CHILD
-        if (execvp(command.c_str(), args.data()) == -1) {
-            perror("Command not found.");
-            exit(1);
+    std::vector<pid_t>& backgroundProcesses) {
+    try {
+        pid_t pid = fork();
+        if (pid == 0) {
+            // CHILD
+            int inputFd = -1;
+            auto inputIt = std::find_if(args.begin(), args.end(),
+                [](const char* arg) { return arg && std::string(arg) == "<"; });
+            if (inputIt != args.end() && std::next(inputIt) != args.end() && *std::next(inputIt)) {
+                inputFd = open(*std::next(inputIt), O_RDONLY);
+                if (inputFd == -1) {
+                    perror("Failed to open input file");
+                    exit(1);
+                }
+                dup2(inputFd, STDIN_FILENO);
+                args.erase(inputIt, std::next(inputIt, 2));
+            }
+
+            int outputFd = -1;
+            auto outputIt = std::find_if(args.begin(), args.end(),
+                [](const char* arg) { return arg && std::string(arg) == ">"; });
+            if (outputIt != args.end() && std::next(outputIt) != args.end() && *std::next(outputIt)) {
+                outputFd = open(*std::next(outputIt), O_WRONLY | O_CREAT | O_TRUNC, 0644);
+                if (outputFd == -1) {
+                    perror("Failed to open output file");
+                    exit(1);
+                }
+                dup2(outputFd, STDOUT_FILENO);
+                args.erase(outputIt, std::next(outputIt, 2));
+            }
+
+            int pipeIdx = -1;
+            for (size_t i = 0; i < args.size(); ++i) {
+                if (args[i] && std::string(args[i]) == "|") {
+                    pipeIdx = i;
+                    break;
+                }
+            }
+
+            if (pipeIdx != -1) {
+                std::vector<char*> leftArgs(args.begin(), args.begin() + pipeIdx);
+                std::vector<char*> rightArgs(args.begin() + pipeIdx + 1, args.end());
+
+                int pipefd[2];
+                if (pipe(pipefd) == -1) {
+                    perror("Pipe failed");
+                    exit(1);
+                }
+
+                pid_t childPid = fork();
+                if (childPid == 0) {
+                    close(pipefd[1]);
+                    dup2(pipefd[0], STDIN_FILENO);
+                    close(pipefd[0]);
+
+                    execvp(rightArgs[0], rightArgs.data());
+                    perror("Command not found.");
+                    exit(1);
+                }
+                else if (childPid > 0) {
```

Comment:
What? Why are you running the left side of the pipe in the current child process?